### PR TITLE
semi-supervised

### LINKED ISSUE #63: Slow indexing
Code has become very slow probably due to an indexing mistake on our side... @maxime1310 and I will try to identify the bottleneck

### PR DESCRIPTION
**Pull request by Edouard360**
* add semi_supervised=False flag in run_benchmarks
* created the VAEC class (VAE for Classification)
* VAE functions signature made compatible with VAEC (label=None)
* Change VAE structure to be more flexible (encoder_l, encoder_z)
* Better names for some VAE functions (get_sample_scale and get_sample_rate distinction)
* Merge loss into the forward function

* Handles both sparse and dense dataset resolves #63

### CODE REVIEW

File: scvi/vae.py [hunk 1]
```diff
@@ -36,64 +36,62 @@ def __init__(self, n_input, n_hidden=128, n_latent=10, n_layers=1,
         if self.dispersion == "gene":
             self.register_buffer('px_r', torch.randn(self.n_input, ))
 
-        self.encoder = Encoder(n_input, n_hidden=n_hidden, n_latent=n_latent, n_layers=n_layers,
-                               dropout_rate=dropout_rate, using_cuda=self.using_cuda)
+        self.z_encoder = Encoder(n_input, n_hidden=n_hidden, n_latent=n_latent, n_layers=n_layers,
+                                 dropout_rate=dropout_rate, using_cuda=self.using_cuda)
+        self.l_encoder = Encoder(n_input, n_hidden=n_hidden, n_latent=1, n_layers=1,
+                                 dropout_rate=dropout_rate, using_cuda=self.using_cuda)
```
Comment (jeff-regier) [inline scvi/vae.py]: The encoders and the decoder shouldn't need to know whether your using cuda or not. Their parameters get transferred to the GPU automatically when you call `vae.cuda()`. Better not to pass them arguments that aren't (or shouldn't) be used.

File: scvi/vae.py [hunk 2]
```diff
@@ -136,33 +132,22 @@ def __init__(self, n_input, n_hidden=128, n_latent=10, n_layers=1, dropout_rate=
...
-        # The process is similar than for encoding q(z/x), except there is always only one hidden layer
-        self.l_encoder = nn.Sequential(
-            nn.Dropout(p=self.dropout_rate),
-            nn.Linear(n_input, n_hidden),
-            nn.BatchNorm1d(n_hidden, eps=1e-3, momentum=0.99),
-            nn.ReLU())
-
-        self.l_mean_encoder = nn.Linear(n_hidden, 1)
-        self.l_var_encoder = nn.Linear(n_hidden, 1)
+    def reparameterize(self, mu, var):
```
Comment (jeff-regier) [inline scvi/vae.py]: With pytorch 0.4, we can delete this `reparameterize` function and use `Normal(mu, var.sqrt()).rsample()` instead. Good to use the built in functions whenever possible so we'll have less code to maintain.

Comment (jeff-regier) [reply scvi/vae.py]: e.g.,
```
>>> from torch.distributions import Normal
>>> Normal(0, 1).rsample()
tensor(-0.9719)
```

File: scvi/vaec.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        kl_divergence = (kl_divergence.t() * proba).sum(dim=1)
+        y_prior = self.y_prior.type(proba.type())
+        kl_divergence += torch.sum(torch.mul(proba, torch.log(y_prior) - torch.log(proba + 1e-8)), dim=-1)
+
+        return reconst_loss, kl_divergence
+
+
+# Classifier
+class Classifier(nn.Module):
+    def __init__(self, n_input, n_hidden=128, n_labels=10, n_layers=1, dropout_rate=0.1, using_cuda=True):
```
Comment (jeff-regier) [inline scvi/vaec.py]: Here too, better not to take `using_cuda` as an argument.

**Comment by jeff-regier** [comment]
Great!
